### ISSUE TITLE
Anti-Aliasing

### ISSUE THREAD

**Issue by Byter09** [ryanisaacg/quicksilver#254]
Hello!

So first of all, I looked through some of the basic but also some major 2D and 3D engines and found this one to be the "cleanest". Then of course, when my decision was clear, I started to test around and maybe switch to another close candidate if a feature is not present.

I looked at some of the engines and noticed, that quicksilver does not mention any anti-aliasing anywhere in the docs, in open or closed issues or anywhere on reddit.

Now, me being the total beginner, I thought "maybe its just a algorithm that you can implement yourself" (and I'd totally do that), but what about all the different kinds of AA methods? `ggez` for example has MSAA, which by itself works but is pretty heavy if you try to render something in 4K with tons of draw calls.

So, I hope I did not overlook something in the docs or maybe I simply do not understand it, but does quicksilver support, or will it support AA in any form or shape? Maybe even per `draw()` / `sprite()` method?

**Comment by ryanisaacg** [comment]
I haven't really looked into AA because I haven't had a need for it. I'm curious, what's your use case?

If there's a great reason to have it I'd put it on my list of things to do, and I'd definitely merge in a PR that added it.

Good luck with picking the right tool!

**Comment by Byter09** [comment]
There's no specific use case for smooth lines, is there?

In my case its pretty... weird. It's a game/demo project I'm working on which will start you in a spaceship around earth. You'll be able to zoom as far as you want out and in - and having a planet next to you, that does not render as a bunch of lines, would be plus. :laughing:

I found the line that determines it for a circle:
https://github.com/ryanisaacg/quicksilver/blob/4910e17f617cbc61ec7370c8f81cb72f8f0dc192/src/graphics/drawable.rs#L187
If you for example increase this on a big circle, it becomes smoother. So allowing to set it would be a start.

Sadly, this is not really AA, as the circle will still draw narrow, one-pixel-wide lines. Same with rectangles or any straight lines. So theoretically (if my rendering classes taught me anything), we would need to tell the underlying back-end to draw each line thats following to be smooth. I did not work much with quicksilver yet but I'm gonna clone it and play around.
For a start, I'll have a look at these: [Anti-aliasing - Wikipedia](https://en.wikipedia.org/wiki/Anti-aliasing)
Then I'd have to remember all the stuff I learned about vertices and how a GPU draws, which by itself will take some time. Hah!


Oh and I choose the right tool already. As I said, quicksilver is the only "clean" engine that I could find, except for the big modular ones (which get ugly if you add everything you need). And once I choose a tool and it misses something, that might benefit everyone - not just my use case - I open an issue. And voilà, here we are :smile:

**Comment by ryanisaacg** [comment]
The inability to draw true lines in quicksilver is more of a historic failure on my part, but basically OpenGL has different drawing "modes" and you can't mix modes within the same draw call. Minimizing draw calls is something quicksilver is designed to do, so it will only draw things in the `GL_TRIANGLES` mode. If I could give access to geometry shaders to users, this could probably be fixed, but I don't want to expose underlying graphics APIs. 

As for how to go about better, smoother triangle versions of shapes, I'm confident that lyon will help with that, so watch #247 for that. (Feel free to help out with it, too, if you like.) It seems like it provides a vector-graphics to triangles version, which should make drawing much nicer shapes with quicksilver a breeze.

**Comment by Byter09** [comment]
I already saw these and looked into Lyon. Looks fantastic! As I said, I'll play around with it and when something useful comes up I'll update this issue again and probably #247. And to stay on topic: Antialiasing can only be implemented per vertice and the triangle it forms with others, right, I'll also look into that.

Oh and I fully support your decision to keep the graphics APIs hidden. OpenGL is a mess (not functionally but the general overview and the huge amounts of boilerplate code to do some tiny thing) and quicksilver is quite tidy. Lets keep it that way 😄

**Comment by Byter09** [comment]
@ryanisaacg I found something, that you have to look into for me, as I have no overview yet how everything works together in quicksilver, and I dont want to produce garbage :D

OpenGL supports per window MSAA by using
```C
glfwWindowHint(GLFW_SAMPLES, 4);
glEnable(GL_MULTISAMPLE);  
```
As far as I know, this is the same method ggez uses, but they used an Enum instead of an int to restricts users from entering garbage sample sizes. So far this would technically solve this issue in less than 2 lines of OpenGL code and probably an additional method and enum in quicksilver itself. I'll try to do this myself but don't count on it.

Tell me what you think and I'll keep looking :)

**Comment by ryanisaacg** [comment]
@Byter09 Would you prefer full-screen antialiasing as a setting or a more granular method? Finally getting around to working on this

**Comment by ryanisaacg** [comment]
You could hypothetically set MSAA for a `Surface`, because you can apply AA to framebuffers. That's about as granular as you can go actually, so a per-drawable approach would probably be too expensive behind the scenes.

I think for an initial implementation full-screen MSAA is a good idea, and then a possible FXAA option could be added later. The problem with FXAA is that it requires custom shader support, and I don't want to add much more complexity to the backend until the gfx backend is in place.

**Comment by Byter09** [comment]
Well then there you go. That's a pretty good start.

And when MSAA is done, a new FXAA specific issue can be set to "blocked" until custom shaders are in. Also maybe that's for milestone 0.4 then? :D

But yeah MSAA alone is quite beautiful, can easily be applied to several of the examples and solves half of all the use cases I mentioned. :P

**Comment by ryanisaacg** [comment]
Closed now that MSAA support is merged, FXAA can be considered bundled with #362
